src/PMGDQueryHandler.cc

```diff
@@ -241,6 +244,50 @@ int PMGDQueryHandler::add_node(const protobufs::AddNode &cn,
     return 0;
 }
 
+int PMGDQueryHandler::update_node(const protobufs::UpdateNode &un,
+                    protobufs::CommandResponse *response)
+{
+    long id = un.identifier();
+    bool query = un.has_query_node();
+
+    // Make sure there is either an id or a query node
+    if (id < 0 && !query) {
+        set_response(response, PMGDCmdResponse::Error, "No way to find update node\n");
+        return -1;
+    }
+
+    auto it = _cached_nodes.find(id);
+    if (it == _cached_nodes.end()) {
+        if (!query) {
+            set_response(response, PMGDCmdResponse::Error, "Undefined _ref value used in update\n");
+            return -1;
+        }
+        else {
+            query_node(un.query_node(), response);
+            if (response->error_code() != PMGDCmdResponse::Success)
+                return -1;
+            it = _cached_nodes.find(un.query_node().identifier());  // id could have been -ve
+        }
+    }
```

Comment:
Perhaps lines 251-271 should be something like this?

```
if (id >= 0) {  
    it = _cached_nodes.find(id);  
    if (it == cached_nodes.end()) {
        set_response(response, PMGDCmdResponse::Error, "Undefined _ref value used in update\n");
        return -1;
    }
}
else if (un.has_query_node()) {
    query_node(un.query_node(), response);
    if (response->error_code()  != PMGDCmdResponse::Success)
        return -1;
    it = _cached_nodes.find(un.query_node().identifier());
}
else {
    set_response(response, PMGDCmdResponse::Error, "No way to find update node\n");
    return -1;
}
```

Reply:
Fixed.